Ported AbstractDescriptor and Descriptor to Kotlin.

toddATavail opened the pull request:
Ported AbstractDescriptor and Descriptor to Kotlin.

- Single commit addresses all discrepancies of previous attempts, and is based on `origin/descriptor_breakup`. This should resolve any incompleteness and incompatibility issues that my previous attempts had. 🤞 
- Upgraded all dependencies to the most recent stable versions.
  - Kotlin -> 1.3.72
  - Shadow -> 5.2.0
  - ASM -> 8.0.1
  - JUnit -> 5.6.2
  - Gradle -> 6.4.1
- Switched all run configurations to use a build JAR instead of a class
  directory. This circumvents issues with the new Kotlin compiler's
  integration into IntelliJ.
- Fixed an issue with building the Avail CLI JAR that caused the
  indexed file analyzer to be run instead of the command-line compiler.
  The command-line compiler is now working again.
- Separated the indexed file analyzer out into its own
  module/subproject.
- Introduced a run JAR configuration for the indexed file analyzer. It
  simply presents the usage message and exits.
- Introduced a script (analyze) for the indexed file analyzer. Verified
  its correct operation.
- Added explicit duplication strategies for Copy tasks and Jar tasks in
  Gradle, in preparation for upcoming 7.0.
- Fixed spacing throughout the project, especially inside KDoc links (by
  translating literal space characters into entity escapes).
- Moved AbstractDescriptor and Descriptor to
  org.avail.descriptor.representation.
- Eliminated @AvailMethod.
- Eliminated Continuation[01234]* and Transformer*.
- Everything runs.
  - All tests pass.
  - All web applications run.
  - The workbench runs.
  - The command-line compiler runs (again).
  - The indexed file analyzer runs.

markATavail commented on .idea/inspectionProfiles/Project_Default.xml:
```diff
@@ -124,6 +124,7 @@
       <option name="m_ignoreLoopsWithoutConditions" value="false" />
     </inspection_tool>
     <inspection_tool class="ForLoopThatDoesntUseLoopVariable" enabled="true" level="WARNING" enabled_by_default="true" />
+    <inspection_tool class="FunctionName" enabled="false" level="WEAK WARNING" enabled_by_default="false" />
```
Which inspection is this?  There doesn't seem to be a way to map between inspections' formal and informal names.

toddATavail replied:
It didn't like `o_*` names. I disabled the inspection, since it totally doesn't work for the project.

markATavail commented on avail-workbench/build.gradle:
```diff
@@ -25,7 +25,7 @@
  * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
  * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
- * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGEN             CE OR OTHERWISE)
```
Fell asleep on the keyboard?  :-)

toddATavail replied:
Huh. No explanation. Looks to be the only occurrence in the code base. Fixed.

markATavail commented on src/main/java/com/avail/descriptor/tuples/package-info.java:
```diff
@@ -30,6 +30,4 @@
  */
 
-@NonnullByDefault
 package com.avail.descriptor.tuples;
```
Did you mean to strip out the non-null default for the tuple package?

toddATavail replied:
Definitely not. The same thing happened for `com.avail.descriptor.types`. I will fix any other occurrences of this also.

markATavail commented on src/main/java/com/avail/descriptor/types/PrimitiveTypeDescriptor.java:
```diff
@@ -317,44 +319,43 @@ public boolean o_IsSupertypeOfPojoBottomType (
+	@Override
 	public boolean o_IsSupertypeOfEnumerationType (
 		final AvailObject object,
 		final A_BasicObject anEnumerationType)
```
I believe this can be strengthened to A_Type.

toddATavail replied:
Requires a surprising number of updates, but yes, you are correct. Fixed.

markATavail commented on src/main/java/com/avail/descriptor/types/package-info.java:
```diff
@@ -30,6 +30,4 @@
  * POSSIBILITY OF SUCH DAMAGE.
  */
 
-@NonnullByDefault
 package com.avail.descriptor.types;
-import com.avail.annotations.NonnullByDefault;
```
Here's the other accidental change you mentioned.

markATavail commented on src/main/java/com/avail/optimizer/L1Translator.java:
```diff
@@ -1373,21 +1447,21 @@ void leafVisit (
 	 *        These become strengthened by type tests in the current manifest.
 	 * @param callSiteHelper
 	 *        The {@link CallSiteHelper} object for this dispatch.
-	\	 */
```
What, did *I* fall asleep on the keyboard?  I'm sorry miss, I must have a cold.

toddATavail replied:
Heh. Thinking about it, probably it was my cat that was so neglegian            t, but yeah, you probably fell asleep. 😄

markATavail commented on src/main/java/com/avail/utility/evaluation/Combinator.java:
```diff
@@ -35,31 +35,14 @@
 
 /**
- * Utility for applying a Continuation1 recursively.
+ * Utility for applying a functions recursively.
```
function

markATavail replied:
Actually, it's a "Utility for allowing a function body to recursively invoke, or otherwise reference the function itself."

toddATavail replied:
Updated the whole message to say:

```
Utilities for applying functions recursively.
```

markATavail commented on src/main/kotlin/com/avail/builder/ModuleNameResolver.kt:
```diff
@@ -41,8 +41,8 @@ import java.util.*
- * [modules][ModuleDescriptor] to [absolute][File.isAbsolute] [file
- * references][File].
+ * [modules][ModuleDescriptor] to [absolute][File.isAbsolute]
+ * [file references][File].
```
&#32;

toddATavail replied:
It posted your comment as nothing at all — no body — but I see what you want fixed, and it's fixed.

toddATavail replied:
FYI. When I edit your comment to see the source, it shows that you typed `&#32;`, but when you don't use code quotes it just posts a space.

markATavail commented on src/main/kotlin/com/avail/compiler/AvailCompiler.kt:
```diff
@@ -4555,9 +4588,9 @@ class AvailCompiler(
 					return
 				}
 				val parts1 = ArrayList<A_Phrase>()
-				phrase1.value.childrenDo(Consumer { parts1.add(it) })
+				phrase1.value.childrenDo { parts1.add(it) }
 				val parts2 = ArrayList<A_Phrase>()
-				phrase2.value.childrenDo(Consumer { parts2.add(it) })
+				phrase2.value.childrenDo { parts2.add(it) }
```
I believe this can be reduced further to ...childrenDo(parts1::add)  (and parts2)

toddATavail replied:
Fixed.

markATavail commented on src/main/kotlin/com/avail/builder/AvailBuilder.kt:
```diff
@@ -88,8 +99,7 @@ import kotlin.collections.set
- * An `AvailBuilder` [compiles][AvailCompiler] and installs into an [Avail
- * runtime][AvailRuntime] a target [module][ModuleDescriptor] and each of its
+ * An `AvailBuilder` [compiles][AvailCompiler] and installs into an [Avail&#32;runtime][AvailRuntime] a target [module][ModuleDescriptor] and each of its
  * dependencies.
```
Long line.

markATavail commented on src/main/kotlin/com/avail/compiler/ParserState.kt:
```diff
@@ -204,8 +203,8 @@ class ParserState internal constructor(
-	 *   A [transformer][Transformer1] that accepts the stringified values and
-	 *   answers an expectation message.
+	 *   A transformer that accepts the stringified values and answers an
+	 *   expectation message.
```
Down a few lines (GitHub won't let me tag them because they're unchanged), `transformer` is the last argument, but it's lined up with no visual separator with `expected`, which makes it hard to tell where the arguments end and the body begins.  I've been advocating hitting return between `String` and `) =` for cases like this.

toddATavail replied:
I reformatted to:

```kt
	internal fun expected(
			level: ParseNotificationLevel,
			values: List<A_BasicObject>,
			transformer: (List<String>)->String) =
		expected(level) { continuation ->
			Interpreter.stringifyThen(
				currentRuntime(),
				lexingState.compilationContext.textInterface,
				values
			) {
				continuation(transformer(it))
			}
		}
```

Objections?

markATavail replied:
If Kotlin won't preserve the double-indented arguments (try cutting and pasting back the whole method, say), then I prefer the uglier form where ") =" goes to a new line.

toddATavail replied:
Ah, you're right. Okay, now it's:

```kt
	internal fun expected(
		level: ParseNotificationLevel,
		values: List<A_BasicObject>,
		transformer: (List<String>)->String
	) =
		expected(level) { continuation ->
			Interpreter.stringifyThen(
				currentRuntime(),
				lexingState.compilationContext.textInterface,
				values
			) {
				continuation(transformer(it))
			}
		}
```

That round-trips across cut and paste.

markATavail commented on src/main/kotlin/com/avail/compiler/problems/Problem.kt:
```diff
@@ -58,9 +58,7 @@ import java.text.MessageFormat
- *   zero-based subscript of the full-Unicode [code
- *   points][CharacterDescriptor] of the file.  [Surrogate
- *   pairs][Character.isSurrogate] are treated as a single code point. It is
+ *   zero-based subscript of the full-Unicode [code&#32;points][CharacterDescriptor] of the file.  [Surrogate&#32;pairs][Character.isSurrogate] are treated as a single code point. It is
```
Long line.

markATavail commented on src/main/kotlin/com/avail/compiler/splitter/ArgumentForMacroOnly.kt:
```diff
@@ -48,8 +50,7 @@ import com.avail.descriptor.types.A_Type
  *
- * Because [list phrases][ListPhraseDescriptor] have an [expression
- * type][A_Phrase.expressionType] that depends on the types of the
+ * Because [list&#32;phrases][ListPhraseDescriptor] have an [expression&#32;type][A_Phrase.expressionType] that depends on the types of the
```
Long line.

markATavail commented on src/main/kotlin/com/avail/compiler/splitter/CaseInsensitive.kt:
```diff
@@ -39,8 +39,7 @@ import com.avail.exceptions.SignatureException
  * `CaseInsensitive` is a special decorator [expression][Expression] that causes
- * the decorated expression's keywords to generate [parse
- * instructions][ParsingOperation] that cause case insensitive parsing. It is
+ * the decorated expression's keywords to generate [parse&#32;instructions][ParsingOperation] that cause case insensitive parsing. It is
```
Long line.

markATavail commented on src/main/kotlin/com/avail/compiler/splitter/Counter.kt:
```diff
@@ -58,8 +58,7 @@ import java.util.*
  *
- * When a double dagger appears in a counter, the counter produces a [whole
- * number][IntegerRangeTypeDescriptor.wholeNumbers] that indicates the number of
+ * When a double dagger appears in a counter, the counter produces a [whole&#32;number][IntegerRangeTypeDescriptor.wholeNumbers] that indicates the number of
```
Long line.

markATavail commented on src/main/kotlin/com/avail/compiler/splitter/MessageSplitter.kt:
```diff
@@ -119,17 +153,16 @@ class MessageSplitter
 	 *  * Placing a [question mark][Metacharacter.QUESTION_MARK]
-	 *    (?) after a group containing arguments but no [double
-	 *    dagger][Metacharacter.DOUBLE_DAGGER] (‡) will limit the repetitions of
+	 *    (?) after a group containing arguments but no [double&#32;dagger][Metacharacter.DOUBLE_DAGGER] (‡) will limit the repetitions of
```
Long line.

toddATavail replied:
There were actually two long lines in this file. I fixed both.

markATavail commented on src/main/kotlin/com/avail/compiler/splitter/Optional.kt:
```diff
@@ -55,8 +58,7 @@ import java.util.*
  * question mark (?).  It may not contain [Argument]s or subgroups, and since it
- * is not a group it may not contain a [double
- * dagger][Metacharacter.DOUBLE_DAGGER] (‡).
+ * is not a group it may not contain a [double&#32;dagger][Metacharacter.DOUBLE_DAGGER] (‡).
```
Long line.  You'll need to break it before "`[double`".

toddATavail replied:
Yep. Reflow does it correctly automatically, and produced the result that you described.

markATavail commented on src/main/kotlin/com/avail/compiler/splitter/RawTokenArgument.kt:
```diff
@@ -39,8 +39,7 @@ import com.avail.descriptor.types.A_Type
  * A `RawTokenArgument` is an occurrence of [ellipsis][Metacharacter.ELLIPSIS]
- * (…) in a message name, followed by an [exclamation
- * mark][Metacharacter.EXCLAMATION_MARK] (!). It indicates where *any* raw token
+ * (…) in a message name, followed by an [exclamation&#32;mark][Metacharacter.EXCLAMATION_MARK] (!). It indicates where *any* raw token
```
Long line.

markATavail commented on src/main/kotlin/com/avail/compiler/splitter/SectionCheckpoint.kt:
```diff
@@ -39,8 +39,7 @@ import com.avail.descriptor.phrases.A_Phrase
 /**
- * An `SectionCheckpoint` expression is an occurrence of the [section
- * sign][Metacharacter.SECTION_SIGN] (§) in a message name.  It indicates a
+ * An `SectionCheckpoint` expression is an occurrence of the [section&#32;sign][Metacharacter.SECTION_SIGN] (§) in a message name.  It indicates a
```
Long line.

markATavail commented on src/main/kotlin/com/avail/descriptor/bundles/MessageBundleDescriptor.kt:
```diff
@@ -291,7 +287,7 @@ class MessageBundleDescriptor private constructor(
 		 *   The affected message bundle.
 		 * @param grammaticalRestriction
-		 *   A [grammatical&#32;restriction][A_GrammaticalRestriction].
+		 *   A [grammatical restriction][A_GrammaticalRestriction].
```
`&#32;`

markATavail commented on src/main/kotlin/com/avail/descriptor/bundles/MessageBundleDescriptor.kt:
```diff
@@ -335,7 +331,7 @@ class MessageBundleDescriptor private constructor(
 		 *   A MessageSplitter for this message name.
 		 * @return
-		 *   A new [message&#32; bundle][A_Bundle].
+		 *   A new [message bundle][A_Bundle].
```
`&#32;` (but without space).

toddATavail replied:
I also fixed a busted instance of this in the class comment.

markATavail commented on src/main/kotlin/com/avail/descriptor/bundles/MessageBundleTreeDescriptor.kt:
```diff
@@ -258,7 +279,7 @@ class MessageBundleTreeDescriptor private constructor(
 		 * stream (so there's no value in checking whether it's an X, whether
 		 * it's a Y, whether it's a Z, etc. Instead, the [LAZY_INCOMPLETE] and
-		 * [LAZY_INCOMPLETE_CASE_INSENSITIVE]&#32;[map][A_Map] takes care of
+		 * [LAZY_INCOMPLETE_CASE_INSENSITIVE] [map][A_Map] takes care of
```
This is correct, but what was there before is interesting!

toddATavail replied:
Yeah, somebody regex'd wrong. 🙂

markATavail commented on src/main/kotlin/com/avail/descriptor/bundles/MessageBundleTreeDescriptor.kt:
```diff
@@ -393,14 +413,14 @@ class MessageBundleTreeDescriptor private constructor(
-				})
-			})
+				}
+			}
```
On the next line, please replace with mutableListOf().

toddATavail replied:
```kt
			val sorted = mutableListOf<String>()
```

markATavail commented on src/main/kotlin/com/avail/descriptor/character/CharacterDescriptor.kt:
```diff
@@ -167,21 +160,19 @@ class CharacterDescriptor private constructor(
 		self: AvailObject,
 		classHint: Class<*>?
-	): Any? {
+	): Any {
```
Ah yes, for this particular class it's never null.

markATavail commented on src/main/kotlin/com/avail/descriptor/fiber/FiberDescriptor.kt:
```diff
@@ -1001,44 +980,26 @@ class FiberDescriptor private constructor(
 		self.mutableSlot(DEBUG_LOG).javaObjectNotNull()
 
-	override fun o_Lock(self: AvailObject, critical: Continuation0) {
+	override fun <T> o_Lock(self: AvailObject, supplier: ()->T): T =
```
Now would be a good time to rename the supplier argument to "body".

markATavail commented on src/main/kotlin/com/avail/descriptor/fiber/FiberDescriptor.kt:
```diff
@@ -1001,44 +980,26 @@ class FiberDescriptor private constructor(
 		}
 
-	override fun mutable(): FiberDescriptor = mutable
+	override fun mutable(): AbstractDescriptor = mutable
```
We can leave off the type declaration.  The val fields are local to the companion, so Kotlin won't complain about "platform operation" result types.

toddATavail replied:
Whoops, yep, thought I got most of these. I'll do a regex right now to fix this systematically, so don't waste your time documenting more of them.

toddATavail replied:
Okay, nailed them. And everything still compiles and passes the tests, so the regex was good. 👍

markATavail commented on src/main/kotlin/com/avail/descriptor/functions/FunctionDescriptor.kt:
```diff
@@ -100,7 +99,7 @@ class FunctionDescriptor private constructor(
 	 */
 	enum class ObjectSlots : ObjectSlotsEnum {
-		/** The [compiled&#32;code][CompiledCodeDescriptor]. */
+		/** The [compiled code][CompiledCodeDescriptor]. */
```
`&#32;`

markATavail commented on src/main/kotlin/com/avail/descriptor/methods/MethodDescriptor.kt:
```diff
@@ -382,13 +413,12 @@ class MethodDescriptor private constructor(
 	override fun o_AddSemanticRestriction(
 		self: AvailObject,
-		restriction: A_SemanticRestriction
+		restrictionSignature: A_SemanticRestriction
```
I think `restriction` was a better name.

toddATavail replied:
Yeah, it definitely was. It was a bit of laziness on my part — I was renaming it to match the base definitions, so that Kotlin wouldn't complain about named parameters. I can patch it up.

markATavail commented on src/main/kotlin/com/avail/descriptor/phrases/SendPhraseDescriptor.kt:
```diff
@@ -245,13 +234,13 @@ class SendPhraseDescriptor private constructor(
 	companion object {
 		/**
-		 * Create a new [send&#32;phrase][SendPhraseDescriptor] from the
+		 * Create a new [send phrase][SendPhraseDescriptor] from the
```
`&#32;`

markATavail commented on src/main/kotlin/com/avail/descriptor/representation/AbstractDescriptor.kt:
```diff
@@ -0,0 +1,4202 @@
+	 * @param name
+	 *   The name of the slot.
+	 */
+	class DebuggerObjectSlots (val name: String) : ObjectSlotsEnumJava
```
If possible, use ObjectSlotsEnum.  ObjectSlotsEnumJava is temporary.

toddATavail replied:
It was not possible.

markATavail commented on src/main/kotlin/com/avail/descriptor/representation/AbstractDescriptor.kt:
```diff
@@ -0,0 +1,4202 @@
+			{
+				null
+			})
+		val fields: MutableList<AvailObjectFieldHelper> = ArrayList()
```
mutableListOf()

markATavail commented on src/main/kotlin/com/avail/descriptor/representation/AbstractDescriptor.kt:
```diff
@@ -0,0 +1,4202 @@
+					HideFieldJustForPrinting::class.java) == null)
+			{
+				newlineTab(builder, indent)
+				val slotName = Nulls.stripNull(slot.fieldName())
```
!!

toddATavail replied:
Turns out I don't even need `!!`.

toddATavail replied:
I will look for `Nulls` usages in Kotlin and remove them systematically.

markATavail commented on src/main/kotlin/com/avail/descriptor/representation/AbstractDescriptor.kt:
```diff
@@ -0,0 +1,4202 @@
+
+	/**
+	 * Answer whether these are acceptable [argument types][TypeDescriptor] for
+	 * invoking a [function][FunctionDescriptor] whose type is the self.
```
-the

toddATavail replied:
There are just three bogus ones in `AbstractDescriptor`. All but one other mention of "the self" are about self types, and are appropriate; the last is part of "the self-rescuing princess", which is also appropriate. I fixed the bad ones.

markATavail commented on src/main/kotlin/com/avail/descriptor/representation/AbstractDescriptor.kt:
```diff
@@ -0,0 +1,4202 @@
+		argValues: List<A_BasicObject>): Boolean
+
+	/**
+	 * Answer whether these are acceptable [argument types][TypeDescriptor] for
```
`&#32;`

markATavail replied:
(and more below)

toddATavail replied:
I fixed these systematically everywhere, which resulted in another 400+ changes. I use a few regular expressions and iterated to a fixed point, manually excluding false positives. I handled all links broken across lines. I reflowed all long comments. I will double check everything before I commit.

markATavail commented on src/main/kotlin/com/avail/descriptor/representation/AbstractDescriptor.kt:
```diff
@@ -0,0 +1,4202 @@
+	 * Add a [definition][DefinitionDescriptor] to the receiver. Causes
+	 * dependent chunks to be invalidated.
+	 *
+	 * Macro signatures and non-macro definitions should not be combined in the
+	 * same method.
```
No longer true.

markATavail commented on src/main/kotlin/com/avail/descriptor/representation/AbstractDescriptor.kt:
```diff
@@ -0,0 +1,4202 @@
+
+	abstract fun o_ParsingInstructions (self: AvailObject): A_Tuple
+
+	// TODO: [TLS] Continue here.
```
Remove.

toddATavail replied:
Heh. Whoops.

markATavail commented on src/main/kotlin/com/avail/descriptor/representation/AbstractDescriptor.kt:
```diff
@@ -0,0 +1,4202 @@
+
+	abstract fun o_StartingLineNumber (self: AvailObject): Int
+
+	/**
```
Remove comment.

markATavail commented on src/main/kotlin/com/avail/descriptor/representation/AbstractDescriptor.kt:
```diff
@@ -0,0 +1,4202 @@
+	abstract fun o_MacroDefinitionsTuple (self: AvailObject): A_Tuple
+
+	/**
+	 * @param self
```
Remove comment.

markATavail commented on src/main/kotlin/com/avail/descriptor/representation/AbstractDescriptor.kt:
```diff
@@ -0,0 +1,4202 @@
+
+	abstract fun o_LexerMethod (self: AvailObject): A_Method
+
+	/**
```
Remove remaining comments of this form.

toddATavail replied:
Done. I searched the whole code base and cleaned up anything like this; there were a few more in Stacks.

markATavail commented on src/main/kotlin/com/avail/descriptor/representation/AbstractDescriptor.kt:
```diff
@@ -0,0 +1,4202 @@
+			}
+
+		/**
+		 * A static cache of mappings from [integer slots][IntegerSlotsEnum] to
```
`&#32;`

markATavail commented on src/main/kotlin/com/avail/descriptor/representation/AbstractDescriptor.kt:
```diff
@@ -0,0 +1,4202 @@
+		 * which isn't much of a penalty since it only affects the default
+		 * object printing mechanism.
+		 */
+		private val bitFieldsCache: MutableMap<
```
```
private val bitFieldsCache =
    mutableMapOf<IntegerSlotsEnum, List<BitField>>()
````

markATavail commented on src/main/kotlin/com/avail/descriptor/representation/AbstractDescriptor.kt:
```diff
@@ -0,0 +1,4202 @@
+		 */
+		fun bitFieldsFor (slot: IntegerSlotsEnum): List<BitField>
+		{
+			bitFieldsLock.readLock().lock()
```
`bitFieldsLock.read {`

(and eliminate the try/finally wrapper)

toddATavail replied:
Done. I took care to `return` from the function, not the `read`. I also looked for this pattern elsewhere; there were no other occurrences in Kotlin.

markATavail commented on src/main/kotlin/com/avail/descriptor/representation/AbstractDescriptor.kt:
```diff
@@ -0,0 +1,4202 @@
+			{
+				bitFieldsLock.readLock().unlock()
+			}
+			bitFieldsLock.writeLock().lock()
```
`return bitFieldsLock.write {`

toddATavail replied:
Done. I fixed this everywhere. There were three occurrences in Kotlin.

markATavail commented on src/main/kotlin/com/avail/descriptor/representation/AbstractDescriptor.kt:
```diff
@@ -0,0 +1,4202 @@
+					// this case, not necessarily an Enum.
+					val lookupMethod = describingClass.getMethod(
+						lookupName, Int::class.javaPrimitiveType)
+					val lookedUp = lookupMethod.invoke(
```
```
when (val lookedUp = lookupMethod.invoke(null, value.toInt())) {
   is IntegerEnumSlotDescriptonEnum? -> ...
   null -> ...
   else -> ...
}
```

toddATavail replied:
Took me forever to figure out why copy and paste didn't work. You spelled it `Descripton` instead of `Description`. 🙂

The code of the enclosing `else` block (not shown in your snippet) now reads thus:

```kt
					val lookupMethod = describingClass.getMethod(
						lookupName, Int::class.javaPrimitiveType)
					when (val lookedUp = lookupMethod.invoke(
						null, value.toInt())) {
						is IntegerEnumSlotDescriptionEnum ->
							append(lookedUp.fieldName())
						else ->
							append("null")
					}
```

This is what you wanted, yes?

markATavail commented on src/main/kotlin/com/avail/descriptor/representation/AbstractDescriptor.kt:
```diff
@@ -0,0 +1,4202 @@
+			}
+		}
+
+		// TODO MvG Comment me
```
```
	/**
	 * Write a description of this [Long] to the builder, taking note that the
	 * value is constrained to contain only numBits of content.  Use conventions
	 * such as grouping into groups of at most four hex digits.
	 *
	 * @param value
	 *   The [Long] to output.
	 * @param numBits
	 *   The number of bits contained in value.
	 * @param builder
	 *   Where to describe the number.
	 */
```

markATavail commented on src/main/kotlin/com/avail/descriptor/representation/AvailObject.kt:
```diff
@@ -350,8 +379,8 @@ class AvailObject private constructor(
 	/**
-	 * Answer whether the [argument&#32;types][AvailObject#argsTupleType]
-	 * supported by the specified [function&#32;type][FunctionTypeDescriptor]
+	 * Answer whether the [argument types][AvailObject#argsTupleType]
```
`&#32;` – there are more places below.

markATavail commented on src/main/kotlin/com/avail/descriptor/representation/BitField.kt:
```diff
@@ -95,8 +95,8 @@ class BitField (
-	 * populated by the default [
-	 * ][AbstractDescriptor.printObjectOnAvoidingIndent] mechanism.
+	 * populated by the default
+	 * [&#32;][AbstractDescriptor.printObjectOnAvoidingIndent] mechanism.
```
`[printing]`

toddATavail replied:
I replaced it with just `[printing]`. Good enough.

markATavail commented on src/main/kotlin/com/avail/descriptor/representation/Descriptor.kt:
```diff
@@ -0,0 +1,2733 @@
+	 * A special enumeration used to visit all object slots within an instance
+	 * of the receiver.
+	 */
+	// Qualified to bypass a Kotlin/Java compiler bug suddenly encountered
+	// by MvG on 2020.04.26.
```
Comment is no longer relevant.

markATavail commented on src/main/kotlin/com/avail/descriptor/representation/Descriptor.kt:
```diff
@@ -0,0 +1,2733 @@
+	}
+
+	override fun o_RepresentationCostOfTupleType (self: AvailObject): Int =
+		// Given two objects that are known to be equal, the second of which is
+		// in the form of a tuple type, is the first one in a better form than
+		// the second one?
+
+		// Explanation: This must be called with a tuple type as the second
+		// argument, but the two arguments must also be equal. All alternative
+		// implementations of tuple types should re-implement this method.
+		unsupportedOperation()
```
Either leave this in braces, or move it into a Kdoc comment.  The code itself doesn't stand out well from the comment, and the paragraph gap makes it twice as hard to read.

toddATavail replied:
I made it a KDoc comment.

markATavail commented on src/main/kotlin/com/avail/descriptor/representation/Descriptor.kt:
```diff
@@ -0,0 +1,2733 @@
+	override fun o_CopyMutablePhrase (self: AvailObject): A_Phrase =
+		unsupportedOperation()
+
+	// Ordinary (non-bin, non-void) objects act as set bins of size one.
```
non-bin, non-nil

toddATavail replied:
Dang, old terminology.

markATavail commented on src/main/kotlin/com/avail/interpreter/levelTwo/operand/L2ReadIntVectorOperand.kt:
```diff
@@ -61,12 +61,12 @@ class L2ReadIntVectorOperand constructor(elements: List<L2ReadIntOperand>)
 
 	override fun clone(
 		replacementElements: List<L2ReadIntOperand>): L2ReadIntVectorOperand =
```
Break before ")".

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/bootstrap/syntax/P_BootstrapInitializingVariableDeclarationMacro.kt:
```diff
@@ -36,24 +36,27 @@ import com.avail.compiler.AvailRejectedParseException
  * The `P_BootstrapInitializingVariableDeclarationMacro` primitive is used for
- * bootstrapping declaration of a [local
- * variable][DeclarationKind.LOCAL_VARIABLE] with an initializing expression.
+ * bootstrapping declaration of a [local&#32;variable][DeclarationKind.LOCAL_VARIABLE] with an initializing expression.
```
Wrap line.

toddATavail replied:
I systematically wrapped everything long earlier today, so I'm going to skip comments like these, assuming that I hit them.

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/bootstrap/syntax/P_BootstrapSuperCastMacro.kt:
```diff
@@ -41,15 +41,18 @@ import com.avail.descriptor.tuples.ObjectTupleDescriptor.tuple
 /**
- * The `P_BootstrapSuperCastMacro` primitive is used to create a [super-cast
- * phrase][SuperCastPhraseDescriptor].  This is used to control method lookup,
+ * The `P_BootstrapSuperCastMacro` primitive is used to create a [super-cast&#32;phrase][SuperCastPhraseDescriptor].  This is used to control method lookup,
```
Wrap line.

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/controlflow/P_CatchException.kt:
```diff
@@ -44,14 +44,19 @@ import com.avail.descriptor.types.TupleTypeDescriptor.zeroOrMoreOf
 /**
- * **Primitive:** Always fail. The Avail failure code invokes the [body
- * block][FunctionDescriptor]. A handler block is only invoked when an exception
+ * **Primitive:** Always fail. The Avail failure code invokes the [body&#32;block][FunctionDescriptor]. A handler block is only invoked when an exception
```
Wrap line.

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/fibers/P_UnparkFiber.kt:
```diff
@@ -44,15 +44,16 @@ import com.avail.descriptor.types.A_Type
 /**
  * **Primitive:** Unpark the specified [fiber][FiberDescriptor]. If the
- * [permit ][SynchronizationFlag.PERMIT_UNAVAILABLE] associated with the fiber
+ * [permit&#32;][SynchronizationFlag.PERMIT_UNAVAILABLE] associated with the fiber
```
Remove spurious `&#32;`.

toddATavail replied:
I cleaned and reflowed all occurrences of this problem. There were three.

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/files/P_CreateDirectory.kt:
```diff
@@ -190,9 +198,9 @@ object P_CreateDirectory : Primitive(5, CanInline, HasSideEffect)
-	 * file permissions][PosixFilePermission].
+	 * Convert the specified [set][A_Set] of
+	 * [ordinals][IntegerDescriptor] into the corresponding [set][Set] of
+	 * [POSIX&#32;file permissions][PosixFilePermission].
```
Add another `&#32;` after "file".

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/files/P_FileRead.kt:
```diff
@@ -70,8 +78,7 @@ import java.nio.channels.AsynchronousFileChannel
 /**
- * **Primitive:** Read the requested number of bytes from the [file
- * channel][AsynchronousFileChannel] associated with the specified
+ * **Primitive:** Read the requested number of bytes from the [file&#32;channel][AsynchronousFileChannel] associated with the specified
```
Wrap line.

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/files/P_FileSetPermissions.kt:
```diff
@@ -132,8 +135,8 @@ object P_FileSetPermissions : Primitive(3, CanInline, HasSideEffect)
-	 * [ordinals][IntegerDescriptor] into the corresponding [set][Set] of [POSIX
-	 * file permissions][PosixFilePermission].
+	 * [ordinals][IntegerDescriptor] into the corresponding [set][Set] of
+	 * [POSIX&#32;file permissions][PosixFilePermission].
```
Add another `&#32;` after "file".

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/general/P_PrintToErrorConsole.kt:
```diff
@@ -52,8 +52,7 @@ import com.avail.io.TextOutputChannel
  * **Primitive:** Print the specified [string][StringDescriptor] to the
- * [current][Interpreter.fiber]'s [standard error channel][TextOutputChannel], [
- * ][ExecutionState.SUSPENDED] the current fiber until the string can be queued
+ * [current][Interpreter.fiber]'s [standard&#32;error&#32;channel][TextOutputChannel], [&#32;][ExecutionState.SUSPENDED] the current fiber until the string can be queued
```
I just read your comment from an earlier file indicating you've reflowed these.  I'm going to continue scanning files within the earlier commit, since I'm at file 536/911.  I'll continue to note needed reflows, and if they're already resolved in your new commit, you can mark them as resolved.

toddATavail replied:
By the way, I found all occurrences of `* ]` and fixed those too. Reflowed as well.

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/general/P_ResetStandardInputStream.kt:
```diff
@@ -40,16 +40,15 @@ import com.avail.descriptor.types.AbstractEnumerationTypeDescriptor.enumerationW
 /**
- * **Primitive:** Reset the [ current fiber][Interpreter.fiber]'s [standard
- * input channel][TextInputChannel] to the previously set
+ * **Primitive:** Reset the [ current fiber][Interpreter.fiber]'s [standard&#32;input channel][TextInputChannel] to the previously set
```
Spurious space, also needs `&#32;`, also reflow.

toddATavail replied:
I handled all of the `**Primitive**` comments specially. I think that they are all fixed now.

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/maps/P_MapTypeKeyType.kt:
```diff
@@ -38,13 +38,14 @@ import com.avail.descriptor.types.InstanceMetaDescriptor.anyMeta
 /**
- * **Primitive:** Answer the key [type][TypeDescriptor] of a [map
- * type][MapTypeDescriptor].
+ * **Primitive:** Answer the key [type][TypeDescriptor] of a [map&#32;type][MapTypeDescriptor].
```
Wrap.

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/maps/P_MapTypeSizes.kt:
```diff
@@ -40,9 +40,11 @@ import com.avail.descriptor.types.IntegerRangeTypeDescriptor
+import com.avail.interpreter.execution.Interpreter
 
 /**
  * **Primitive:** Answer the [size range][IntegerRangeTypeDescriptor] of a [ map
```
`&#32;`

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/maps/P_MapTypeValueType.kt:
```diff
@@ -38,13 +38,14 @@ import com.avail.descriptor.types.InstanceMetaDescriptor.anyMeta
 /**
- * **Primitive:** Answer the value [type][TypeDescriptor] of a [map
- * type][MapTypeDescriptor].
+ * **Primitive:** Answer the value [type][TypeDescriptor] of a [map&#32;type][MapTypeDescriptor].
```
Wrap.

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/methods/P_BundleMessageParts.kt:
```diff
@@ -40,9 +40,11 @@ import com.avail.descriptor.types.FunctionTypeDescriptor.functionType
+import com.avail.interpreter.execution.Interpreter
 
 /**
  * **Primitive:** Answer a [message bundle's][MessageBundleDescriptor] message
```
`&#32;`

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/methods/P_BundleMessage.kt:
```diff
@@ -39,9 +39,11 @@ import com.avail.descriptor.types.A_Type
+import com.avail.interpreter.execution.Interpreter
 
 /**
  * **Primitive:** Answer a [message bundle's][MessageBundleDescriptor] message
```
`&#32;`

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/modules/P_DeclareAllExportedAtoms.kt:
```diff
@@ -46,15 +46,17 @@ import com.avail.descriptor.types.IntegerRangeTypeDescriptor.wholeNumbers
  * deserialize correctly. It forces the given set of atoms to be included in the
- * current module's [public names][ObjectSlots.IMPORTED_NAMES] or [private
- * names][ObjectSlots.PRIVATE_NAMES], depending on the value of the supplied
+ * current module's [public names][ObjectSlots.IMPORTED_NAMES] or [private&#32;names][ObjectSlots.PRIVATE_NAMES], depending on the value of the supplied
```
Wrap.  This isn't wrapped yet in your updated commit.

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/modules/P_VisibleAtoms.kt:
```diff
@@ -43,11 +43,11 @@ import com.avail.descriptor.types.IntegerRangeTypeDescriptor.wholeNumbers
+import com.avail.interpreter.execution.Interpreter
 
 /**
  * **Primitive**: Answer every [true name][A_Atom] visible in the
```
`&#32;`

toddATavail replied:
I found a more aggressive set of regular expressions for finding 1) spaces in links and 2) long lines. I iterated them until stasis. I visited every long line and reflowed. I found a couple hundred more sites and patched them. I can't find any other spaces in links or long comment lines at this point.

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/objects/P_MapToObjectType.kt:
```diff
@@ -45,13 +45,14 @@ import com.avail.descriptor.types.IntegerRangeTypeDescriptor.wholeNumbers
 /**
- * **Primitive:** Convert a [map][MapDescriptor] from fields ([instance
- * types][InstanceTypeDescriptor] of [atoms][AtomDescriptor]) to
+ * **Primitive:** Convert a [map][MapDescriptor] from fields ([instance&#32;types][InstanceTypeDescriptor] of [atoms][AtomDescriptor]) to
```
Wrap.

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/phrases/P_CreateBlockExpression.kt:
```diff
@@ -46,21 +46,23 @@ import com.avail.descriptor.types.FunctionTypeDescriptor.functionType
 /**
- * **Primitive:** Create a []block&#32;expression][BlockPhraseDescriptor] from
- * the specified [argument&#32;declarations][PhraseKind.ARGUMENT_PHRASE],
+ * **Primitive:** Create a [block  expression][BlockPhraseDescriptor] from
```
These links are still not right in the newer version.

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/phrases/P_CreateLiteralExpression.kt:
```diff
@@ -40,13 +40,15 @@ import com.avail.descriptor.types.A_Type
- * **Primitive:** Create a [literal&#32;expression][LiteralPhraseDescriptor]
- * from the specified [literal&#32;token][LiteralTokenDescriptor].
+ * **Primitive:** Create a [literal expression][LiteralPhraseDescriptor]
+ * from the specified [literal token][LiteralTokenDescriptor].
```
`&#32;`

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/phrases/P_CreateRestrictedSendExpression.kt:
```diff
@@ -77,7 +80,7 @@ import java.util.concurrent.atomic.AtomicInteger
  * **Primitive CreateRestrictedSendExpression**: Create a
  * [send&#32;phrase][SendPhraseDescriptor] from the specified
- * [message&#32;bundle][A_Bundle], [list&#32;phrase][ListPhraseDescriptor] of
+ * [message&#32;bundle][A_Bundle], [list phrase][ListPhraseDescriptor] of
```
`[list&#32;phrase]`

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/phrases/P_LiteralExpressionToken.kt:
```diff
@@ -40,13 +40,15 @@ import com.avail.descriptor.types.A_Type
- * **Primitive:** Answer the [literal&#32;token][LiteralTokenDescriptor] from
- * which the specified [literal&#32;phrase][LiteralPhraseDescriptor] was built.
+ * **Primitive:** Answer the [literal token][LiteralTokenDescriptor] from
+ * which the specified [literal phrase][LiteralPhraseDescriptor] was built.
```
The link on the previous line is good, but not this one.

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/phrases/P_PermutedListList.kt:
```diff
@@ -40,13 +40,15 @@ import com.avail.descriptor.types.A_Type
 /**
- * **Primitive**: Answer the [list&#32;phrase][ListPhraseDescriptor] permuted by
- * the specified [permuted&#32;list&#32;phrase][PermutedListPhraseDescriptor].
+ * **Primitive**: Answer the [list phrase][ListPhraseDescriptor] permuted by
```
`&#32;` on both lines.

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/phrases/P_PermutedListPermutation.kt:
```diff
@@ -41,12 +41,14 @@ import com.avail.descriptor.types.FunctionTypeDescriptor.functionType
 
 /**
- * **Primitive**: Answer the [permutation&#32;tuple][TupleDescriptor] of the
+ * **Primitive**: Answer the [permutation tuple][TupleDescriptor] of the
```
`&#32;`

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/phrases/P_SendReturnType.kt:
```diff
@@ -39,12 +39,14 @@ import com.avail.descriptor.types.FunctionTypeDescriptor.functionType
 /**
- * **Primitive:** Answer the specified [send&#32;phrase][SendPhraseDescriptor]'s
+ * **Primitive:** Answer the specified [send phrase][SendPhraseDescriptor]'s
  * expected return []type][TypeDescriptor].
```
Spurious bracket.

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/phrases/P_VariableUseDeclaration.kt:
```diff
@@ -40,14 +40,16 @@ import com.avail.descriptor.types.A_Type
  * **Primitive:** Answer the [declaration][DeclarationPhraseDescriptor] that
  * created the entity referenced
- * by this [variable&#32;use][VariableUsePhraseDescriptor].
+ * by this [variable use][VariableUsePhraseDescriptor].
```
`&#32;` and rewrap previous line.

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/pojos/P_CreatePojoArrayType.kt:
```diff
@@ -42,12 +42,14 @@ import com.avail.descriptor.types.PojoTypeDescriptor
 
 /**
- * **Primitive:** Create a [pojo&#32;array&#32;type][PojoTypeDescriptor] for the
+ * **Primitive:** Create a [pojo array type][PojoTypeDescriptor] for the
```
One is fixed, but not the other.

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/pojos/P_CreatePojoType.kt:
```diff
@@ -44,15 +44,15 @@ import com.avail.descriptor.types.TupleTypeDescriptor.zeroOrMoreOf
  * **Primitive:** Answer the [type][A_Type] that represents the Java [Class]
  * specified by the given fully-qualified name and type parameters. The result
- * is either an Avail [type][A_Type] or a [pojo&#32;type][PojoTypeDescriptor].
+ * is either an Avail [type][A_Type] or a [pojo type][PojoTypeDescriptor].
```
`&#32;`

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/privatehelpers/P_PushConstant.kt:
```diff
@@ -37,17 +37,20 @@ import com.avail.descriptor.types.A_Type
 /**
  * **Primitive:** The first literal is being returned. Extract the first literal
- * from the [compiled&#32;code][CompiledCodeDescriptor] that the interpreter has
+ * from the [compiled code][CompiledCodeDescriptor] that the interpreter has
```
`&#32;`

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/rawfunctions/P_CompiledCodeNumStackSlots.kt:
```diff
@@ -38,13 +38,15 @@ import com.avail.descriptor.types.A_Type
 /**
  * **Primitive:** Answer the number of stack slots (not counting arguments and
- * locals) created for the [compiled&#32;code][CompiledCodeDescriptor].
+ * locals) created for the [compiled code][CompiledCodeDescriptor].
```
`&#32;`

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/sockets/P_ServerSocketIPv6Bind.kt:
```diff
@@ -63,7 +68,7 @@ import java.nio.channels.AsynchronousServerSocketChannel
 /**
  * **Primitive:** Bind the [AsynchronousServerSocketChannel] referenced by the
- * specified [handle][AtomDescriptor] to an [IPv6&#32;address][Inet6Address] and
+ * specified [handle][AtomDescriptor] to an [IPv6 address][Inet6Address] and
```
`&#32;`

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/sockets/P_SocketIPv4Bind.kt:
```diff
@@ -62,7 +67,7 @@ import java.nio.channels.AsynchronousSocketChannel
 /**
  * **Primitive:** Bind the [AsynchronousSocketChannel] referenced by the
- * specified [handle][AtomDescriptor] to an [IPv4&#32;address][Inet4Address] and
+ * specified [handle][AtomDescriptor] to an [IPv4 address][Inet4Address] and
```
`&#32;`

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/sockets/P_SocketIPv4Connect.kt:
```diff
@@ -73,7 +79,7 @@ import java.util.Collections.emptyList
 /**
  * **Primitive:** Connect the [AsynchronousSocketChannel] referenced by the
- * specified [handle][A_Atom] to an [IPv4&#32;address][Inet4Address] and
+ * specified [handle][A_Atom] to an [IPv4 address][Inet4Address] and
```
`&#32;`

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/sockets/P_SocketIPv6Bind.kt:
```diff
@@ -62,7 +67,7 @@ import java.nio.channels.AsynchronousSocketChannel
 /**
  * **Primitive:** Bind the [AsynchronousSocketChannel] referenced by the
- * specified [handle][AtomDescriptor] to an [IPv6&#32;address][Inet6Address] and
+ * specified [handle][AtomDescriptor] to an [IPv6 address][Inet6Address] and
```
`&#32;`

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/sockets/P_SocketIPv6Connect.kt:
```diff
@@ -74,7 +80,7 @@ import java.util.Collections.emptyList
 /**
  * **Primitive:** Connect the [AsynchronousSocketChannel] referenced by the
- * specified [handle][A_Atom] to an [IPv4&#32;address][Inet4Address] and
+ * specified [handle][A_Atom] to an [IPv4 address][Inet4Address] and
```
`&#32;`

toddATavail replied:
This comment should also be able `IPv6Address`. I fixed that too.

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/tuples/P_TupleTypeAtThrough.kt:
```diff
@@ -43,14 +43,15 @@ import com.avail.descriptor.types.IntegerRangeTypeDescriptor.naturalNumbers
  * **Primitive:** Answer the [type][TypeDescriptor] that is the union of the
- * types within the given range of indices of the given [tuple
- * type][TupleTypeDescriptor]. Answer [bottom][BottomTypeDescriptor] if all the
+ * types within the given range of indices of the given [tuple&#32;type][TupleTypeDescriptor]. Answer [bottom][BottomTypeDescriptor] if all the
```
Wrap.

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/tuples/P_TupleTypeConcatenate.kt:
```diff
@@ -38,14 +38,15 @@ import com.avail.descriptor.types.FunctionTypeDescriptor.functionType
  * **Primitive:** Answer the [type][TypeDescriptor] that is the type of all
- * possible concatenations of instances of the given [tuple
- * types][TupleTypeDescriptor]. This is basically the semantic restriction of
+ * possible concatenations of instances of the given [tuple&#32;types][TupleTypeDescriptor]. This is basically the semantic restriction of
```
Wrap.

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/tuples/P_TupleTypeLeadingTypes.kt:
```diff
@@ -40,14 +40,15 @@ import com.avail.descriptor.types.TupleTypeDescriptor
  * **Primitive:** Answer the [tuple][TupleDescriptor] of leading
- * [types][TypeDescriptor] that constrain this [tuple
- * type][TupleTypeDescriptor].
+ * [types][TypeDescriptor] that constrain this [tuple&#32;type][TupleTypeDescriptor].
```
Looks like the `&#32;` pushed this line a bit too long – please wrap.

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/variables/P_DisableTraceVariableReadsBeforeWrites.kt:
```diff
@@ -48,15 +48,13 @@ import com.avail.descriptor.types.TypeDescriptor.Types.TOP
- * **Primitive:** Disable variable [read-before-write
- * tracing][TraceFlag.TRACE_VARIABLE_READS_BEFORE_WRITES] for the [current
- * fiber][FiberDescriptor.currentFiber]. To each [variable][A_Variable] that
+ * **Primitive:** Disable variable [read-before-write&#32;tracing][TraceFlag.TRACE_VARIABLE_READS_BEFORE_WRITES] for the [current&#32;fiber][FiberDescriptor.currentFiber]. To each [variable][A_Variable] that
```
Wrap.

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/variables/P_DisableTraceVariableWrites.kt:
```diff
@@ -45,17 +45,16 @@ import com.avail.descriptor.types.SetTypeDescriptor.setTypeForSizesContentType
+import com.avail.interpreter.execution.Interpreter
 
 /**
  * **Primitive:** Disable [variable write tracing
```
Rewrap this comment.

toddATavail replied:
I also fixed the broken link visible here on line 54.

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/variables/P_GetClearing.kt:
```diff
@@ -41,12 +41,16 @@ import com.avail.descriptor.types.FunctionTypeDescriptor.functionType
+import com.avail.interpreter.execution.Interpreter
 
 /**
  * **Primitive:** Get the value of the [variable][A_Variable], clear the variable, then answer the
```
Rewrap.

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/variables/P_VariableWriteType.kt:
```diff
@@ -37,9 +37,11 @@ import com.avail.descriptor.types.FunctionTypeDescriptor.functionType
+import com.avail.interpreter.execution.Interpreter
 
 /**
  * **Primitive:** Extract the write type of a [variable type ][VariableTypeDescriptor].
```
The latest version has a spurious `&#32;` at the end of the link text.

markATavail commented on src/main/kotlin/com/avail/serialization/SerializerOperation.kt:
```diff
@@ -833,8 +887,9 @@ enum class SerializerOperation constructor(
 	/**
-	 * A [tuple of characters with code points in][StringDescriptor].  Write the
-	 * size of the tuple then the sequence of character bytes.
+	 * A [tuple&#32;of&#32;characters&#32;with code points
```
This is fixed in the newer commit, but `[StringDescriptor]` has a new typo.

markATavail commented on src/main/kotlin/com/avail/serialization/SerializerOperation.kt:
```diff
@@ -902,8 +958,8 @@ enum class SerializerOperation constructor(
 	/**
-	 * A [tuple of integers][TupleDescriptor] whose values all fall in the range
-	 * 0..2^31-1.
+	 * A [tuple&#32;of&#32;integers][TupleDescriptor] whose values all fall in
```
I think just the word "tuple" should be in the link, and "of integers" should be after the link.

markATavail commented on src/main/kotlin/com/avail/tools/options/ValidationException.kt:
```diff
@@ -34,8 +34,7 @@ package com.avail.tools.options
  * Exception thrown by the [factory][OptionProcessorFactory]'s validation
- * process in the event that the client-specified [option
- * processor][OptionProcessor] fails validation.
+ * process in the event that the client-specified [option&#32;processor][OptionProcessor] fails validation.
```
Wrap.

markATavail commented on src/main/kotlin/com/avail/utility/fsm/InvalidContextException.kt:
```diff
@@ -33,8 +33,7 @@
 /**
- * Exception thrown during a state transition when an invalid [execution
- * context][ExecutionContext] is detected
+ * Exception thrown during a state transition when an invalid [execution&#32;context][ExecutionContext] is detected
```
Wrap.

markATavail commented on src/main/kotlin/com/avail/utility/fsm/ValidationException.kt:
```diff
@@ -45,8 +45,7 @@ package com.avail.utility.fsm
  *   A (hopefully) informative message explaining why the
- *   [factory][StateMachineFactory] could not validate the specified [state
- *   machine][StateMachine].
+ *   [factory][StateMachineFactory] could not validate the specified [state&#32;machine][StateMachine].
```
Wrap.

markATavail commented on unicode-catalog/src/main/kotlin/com/avail/tools/unicode/Catalog.kt:
```diff
@@ -344,8 +350,8 @@ internal class Catalog
-	 * A [JSON-friendly representative][JSONFriendly] of the [complete set of
-	 * non-ASCII code][allNonAsciiCodePoints].
+	 * A [JSON-friendly&#32;representative][JSONFriendly] of the
+	 * [complete&#32;set&#32;of&#32;non-ASCII&#32;code][allNonAsciiCodePoints].
```
...`code&#32;points]`

markATavail commented on unicode-catalog/src/main/kotlin/com/avail/tools/unicode/Catalog.kt:
```diff
@@ -368,7 +374,7 @@ internal class Catalog
 	/**
 	 * Answer a [JSON-friendly representative][JSONFriendly] of the
-	 * [complete set of non-ASCII,][allSymbolicCodePoints].
+	 * [complete&#32;set&#32;of&#32;non-ASCII,][allSymbolicCodePoints].
```
...`of the [set][allSymbolicCodePoints] of code points that are neither ASCII nor alpha-numeric.`

toddATavail commented:
@markATavail, I have addressed all issues discovered during code review, and systematically fixed a number of issues related to wrapping and spaces that were not discovered during review. Thanks for putting so much time into this. My latest changes are up.

@richATavail, you may want to rebase your optimizer port onto the latest version of this MR (0a54f22). Most of my changes were to previously translated Kotlin files, so while the change volume is high, the conflict volume should be low. 🤞

markATavail commented on avail-cli/src/main/kotlin/com/avail/tools/compiler/configuration/EnvironmentConfigurator.kt:
```diff
@@ -56,7 +56,7 @@ import com.avail.utility.configuration.Configurator
  *    directory containing source modules, and may be sometimes be omitted
  *    (e.g., when compilation is not required).
- *  * **AVAIL_RENAMES**: The path to the [renames file][RenamesFileParser].
+ *  * **AVAIL_RENAMES**: The path to the [renames&#32;file][RenamesFileParser].
```
Insert blank line before, to avoid maintenance problems with lists.

markATavail commented on avail-server/src/main/kotlin/com/avail/server/configuration/EnvironmentConfigurator.kt:
```diff
@@ -52,10 +52,11 @@ import com.avail.utility.configuration.Configurator
+ *  the absolute path of a binary module repository. `sourceDirectory`
+ *  represents the absolute path of a package, i.e., a directory containing
+ *  source modules, and may be sometimes be omitted (e.g., when compilation is
+ *  not required). * **AVAIL_RENAMES**: The path to the
```
Move "* **" to the a new line, and leave a blank line before it to avoid maintenance problems.

markATavail commented on src/main/kotlin/com/avail/builder/ModuleName.kt:
```diff
@@ -37,9 +37,10 @@ import com.avail.descriptor.module.ModuleDescriptor
 /**
  * A `ModuleName` represents the canonical name of an Avail
  * [module][ModuleDescriptor]. A canonical name is specified relative to an
- * Avail [module&#32;root][ModuleRoots] and has the form R/X/Y/Z**, where **R** is a
```
Add `**` markup before "R/...".  The line will then have to be rewrapped.

markATavail commented on src/main/kotlin/com/avail/builder/ModuleName.kt:
```diff
@@ -37,9 +37,10 @@ import com.avail.descriptor.module.ModuleDescriptor
- * Avail [module&#32;root][ModuleRoots] and has the form R/X/Y/Z**, where **R** is a
- * module root on the Avail module path, **X** is a package within **R**, **Y**
- * is a package within **X**, and **Z** is a module or package within **Y**.
+ * Avail [module&#32;root][ModuleRoots] and has the form R/X/Y/Z**, where **R**
```
Add `**` markup before "R/...".  The line will then have to be rewrapped.

markATavail commented on src/main/kotlin/com/avail/compiler/instruction/AvailLabel.kt:
```diff
@@ -39,7 +39,7 @@ import com.avail.descriptor.tuples.A_Tuple
 
 /**
- * An `AvailLabel` is a pseudo-instruction in the [Level One
+ * An `AvailLabel` is a pseudo-instruction in the [Level&#32;One
```
Missed one.

markATavail commented on src/main/kotlin/com/avail/descriptor/functions/CompiledCodeDescriptor.kt:
```diff
@@ -1005,7 +1005,7 @@ class CompiledCodeDescriptor private constructor(
 			A_Tuple,
 			A_Phrase"""))
-	override fun mutable(): AbstractDescriptor = throw unsupportedOperation()
+	override fun mutable() = throw unsupportedOperation()
```
It looks like `throw unsupportedOperation()` does the same thing as simply `unsupportedOperation()`, and that the `Nothing` type allows the first form to continue to be type correct.  After Kotlinization is complete, we should change them all to the latter form.  We should also remove `unsupportedOperationException()`, and move `unsupportedOperation` into a companion of `AvailUnsupportedOperationException`.  For now, let's just let it be.

toddATavail replied:
I systematically cleaned these.

markATavail commented on src/main/kotlin/com/avail/descriptor/methods/MacroDefinitionDescriptor.kt:
```diff
@@ -6,13 +6,13 @@
+ *    this list of conditions and the following disclaimer in the documentation
  *    and/or other materials provided with the distribution.
- *
+*
```
Some of these copyrights are now mangled on this line.  Don't fix it on this pass, but eventually we should blindly replace anything that starts/ends the same as our copyright with the canonical form (and 1993-2020).

toddATavail replied:
They are all fixed now.

markATavail commented on src/main/kotlin/com/avail/interpreter/levelOne/L1InstructionWriter.kt:
```diff
@@ -61,7 +61,7 @@ import java.util.*
 
 /**
- * An instance of this class can be used to construct a [compiled code
+ * An instance of this class can be used to construct a [compiled&#32;code
```
`&#32;`

markATavail commented on src/main/kotlin/com/avail/interpreter/levelTwo/L2Chunk.kt:
```diff
@@ -74,9 +74,10 @@ import java.util.concurrent.locks.ReentrantLock
 
 /**
- * A Level Two chunk represents an optimized implementation of a [compiled code
+ * A Level Two chunk represents an optimized implementation of a [compiled&#32;code
```
`&#32;`

markATavail commented on src/main/kotlin/com/avail/interpreter/levelTwo/L2Instruction.kt:
```diff
@@ -60,7 +60,7 @@ import java.util.*
  * integers, operation followed by operands.  That is no longer the case, as of
  * 2013-05-01 `MvG`.  Instead, the L2Instructions themselves are kept around for
- * reoptimization and [JVM code generation][translateToJVM].
+ * reoptimization and [JVM&#32;code generation][translateToJVM].
```
`&#32;`

markATavail commented on src/main/kotlin/com/avail/interpreter/levelTwo/L2Operation.kt:
```diff
@@ -66,7 +66,7 @@ import org.objectweb.asm.MethodVisitor
 
 /**
- * The instruction set for the [level two Avail interpreter][Interpreter].
+ * The instruction set for the [level&#32;two Avail interpreter][Interpreter].
```
`&#32;` x 2

markATavail commented on src/main/kotlin/com/avail/interpreter/levelTwo/operation/L2_GET_IMPLICIT_OBSERVE_FUNCTION.kt:
```diff
@@ -45,8 +45,9 @@ import com.avail.optimizer.jvm.JVMTranslator
 /**
- * Store the [implicit observe function][AvailRuntime.implicitObserveFunction]
- * into the supplied [object register][L2BoxedRegister].
+ * Store the [implicit&#32;observe
```
`&#32;`

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/files/P_FileGetPermissions.kt:
```diff
@@ -68,7 +68,7 @@ import java.util.*
 /**
  * **Primitive:** Answer the [ordinals][IntegerDescriptor] (into
- * [IOSystem.posixPermissions]) of the [POSIX file
+ * [IOSystem.posixPermissions]) of the [POSIX&#32;file
```
`&#32;`

markATavail commented on src/main/kotlin/com/avail/interpreter/primitive/methods/P_BundleSignatures.kt:
```diff
@@ -50,7 +50,7 @@ import com.avail.interpreter.execution.Interpreter
 /**
  * **Primitive:** Answer a [set][SetDescriptor] of all currently defined
- * [definitions][DefinitionDescriptor] for the [true message
+ * [definitions][DefinitionDescriptor] for the [true&#32;message
```
`&#32;`
